ComponentKit/Core/Scope/CKComponentScopeRoot.mm

```diff
@@ -9,126 +9,109 @@
+  std::vector<CKComponentScopePredicate> _componentPredicates;
+  std::vector<CKComponentControllerScopePredicate> _componentControllerPredicates;
+  
+  _CKRegisteredComponentsMap _registeredComponents;
+  _CKRegisteredComponentControllerMap _registeredComponentControllers;
 }
 
 + (instancetype)rootWithListener:(id<CKComponentStateListener>)listener
+             componentPredicates:(const std::vector<CKComponentScopePredicate> &)componentPredicates
+   componentControllerPredicates:(const std::vector<CKComponentControllerScopePredicate> &)componentControllerPredicates
 {
   static int32_t nextGlobalIdentifier = 0;
-  return [[CKComponentScopeRoot alloc] initWithListener:listener globalIdentifier:OSAtomicIncrement32(&nextGlobalIdentifier)];
+  return [[CKComponentScopeRoot alloc] initWithListener:listener
+                                       globalIdentifier:OSAtomicIncrement32(&nextGlobalIdentifier)
+                                    componentPredicates:componentPredicates
+                          componentControllerPredicates:componentControllerPredicates];
 }
 
 - (instancetype)newRoot
 {
-  return [[CKComponentScopeRoot alloc] initWithListener:_listener globalIdentifier:_globalIdentifier];
+  return [[CKComponentScopeRoot alloc] initWithListener:_listener
+                                       globalIdentifier:_globalIdentifier
+                                    componentPredicates:_componentPredicates
+                          componentControllerPredicates:_componentControllerPredicates];
 }
 
 - (instancetype)initWithListener:(id<CKComponentStateListener>)listener
                 globalIdentifier:(CKComponentScopeRootIdentifier)globalIdentifier
+             componentPredicates:(const std::vector<CKComponentScopePredicate> &)componentPredicates
+   componentControllerPredicates:(const std::vector<CKComponentControllerScopePredicate> &)componentControllerPredicates
 {
   if (self = [super init]) {
     _listener = listener;
     _globalIdentifier = globalIdentifier;
     _rootFrame = [[CKComponentScopeFrame alloc] initWithHandle:nil];
-    _boundsAnimationComponents = [NSHashTable weakObjectsHashTable];
+    _componentPredicates = componentPredicates;
+    _componentControllerPredicates = componentControllerPredicates;
   }
   return self;
 }
 
-- (void)registerAnnounceableEventsForController:(CKComponentController *)controller
+- (void)registerComponent:(id<CKScopedComponent>)component
 {
-  for (const auto &announceableEvent : announceableEvents()) {
-    if (CKSubclassOverridesSelector([CKComponentController class], [controller class], announceableEvent.second)) {
-      _eventRegistration.insert({{announceableEvent.first, controller}});
+  if (!component) {
+    // Handle this gracefully so we don't have a bunch of nils being passed to predicates.
+    return;
+  }
+  for (const auto &predicate : _componentPredicates) {
+    if (predicate(component)) {
+      _registeredComponents.insert({predicate, component});
```

Comment:
lol yeah, this was supposed to be a multimap, not a straight unordered_map. Good find.